### ISSUE TITLE
Customizable keycode behavior

### ISSUE THREAD

**Issue by MixedMatched** [HaoboGu/rmk#149]
It might be helpful to add a custom handler for key input so that a user can have custom actions that can't be done by existing keycodes. QMK does this with [process_record_user](https://docs.qmk.fm/custom_quantum_functions).

This could probably be implemented by having the user supply an optional function pointer with the signature `KeyAction -> bool` that would be run before the match statement in `Keyboard::process_key_change`. The returned boolean would then control whether the normal action handler is run after the custom handler. This would maybe also require exposing other fields or channels so that the function could add keycodes to be processed or trigger other parts of the firmware.

**Comment by HaoboGu** [comment]
I agree, we should have this

**Comment by tib888** [comment]
How is this related to the ZMK "behavior-mod-morph"?

//this is used to enter "á" or "Á" depending on the shift's state:
hun_aaAA_: hun_aaAA_ {
            compatible = "zmk,behavior-mod-morph";
            #binding-cells = <0>;
            bindings = <&hun_aa>, <&hun_AA_>;
            mods = <(MOD_LSFT|MOD_RSFT)>;
        };

Or Kanata "fork" (this config is MUCH nicer BTW!)?

áÁ (fork (unicode á) (unicode Á) (lsft rsft))

If this issue is about something else then I'll add a separate issue to support such a state-fork behavior simirarly to Combos.
(best would be to be able to fork depending not only based on modifier states, but also based on XXXlock active states and maybe layer activation states).

I guess VIAL has no way to configure it, but I still need it.

**Comment by Exidex** [comment]
I have been looking to migrate my [custom keymap](https://github.com/Exidex/qmk_firmware/blob/custom/keyboards/beekeeb/piantor_pro) to RMK for more than a year now. In QMK, my keymap only uses `process_record_user`, `layer_state_set_user` and custom keycodes, with programmatic key and layer activations/deactivations.

From what I could find there seems to be functionality to listen on the events but not to intercept (allowing to override default behavior). Also almost all of the methods on `KeyMap` are `pub(crate)`, so it is not possible to programmatically press or release keys for crate consumers. Is that intended? To mine uninitiated eye, exposing these 2 things could be enough to cover most of customization use cases, mine included. 

Also it looks like Keyboard runnable includes a lot of functionality beyond the base keyboard handling (stuff like fork, osm, osl, caps_word, etc), which might make override behavior unreliable/flaky and more pluggable infrastructure would make sense?

**Comment by HaoboGu** [comment]
> From what I could find there seems to be functionality to listen on the events but not to intercept (allowing to override default behavior). Also almost all of the methods on `KeyMap` are `pub(crate)`, so it is not possible to programmatically press or release keys for crate consumers. Is that intended? To mine uninitiated eye, exposing these 2 things could be enough to cover most of customization use cases, mine included.
> 
> Also it looks like Keyboard runnable includes a lot of functionality beyond the base keyboard handling, which might make override behavior unreliable/flaky and more pluggable infrastructure would make sense?

Yes, the current design SHOULD expose a way to override the key processing logic. But the current key processing logic is layered and quite complex, it's not implemented yet. The easiest way is to intercept the key event channel, but this requires you to implement the FULL key processing logic, which makes the implementation quite cumbersome. I have no good idea for this right now, if you have any ideas, please open an issue!

**Comment by Exidex** [comment]
I have been thinking about it for some time. So to start I'll try to verbalize what we may need for extension system in general. Mine use case is pretty minimal, but I'll try to be comprehensive.

We have concurrency via Runnables, which means we need some kind of messaging system for extensions and this is what current event system is:
- Event
	- the data that is sent between Runnables
- Processors
	- receive events
	- execute code on specific time (deadline)
	- execute code periodically (polling)
- Input devices
	- read hardware and produce events
- publish_event fn
	- sent any event(s) from anywhere

It only allows for observing the action.

I will go one by one to see which functionality each feature would need to be supported by extension system:
- BLE Passkey Entry
	- key events are swallowed 
- OSM
	- key events are propagated immediately 
	- next key or timeout
	- listens on layer change events
	- manipulate modifiers
- OSL
	- key events are propagated immediately 
	- manipulate layers
	- next key or timeout
- Morse
	- key events are delayed (via buffer_keys)
		- cancel key press/event
		- send hid/event
	- timeout
	- registers/unregisters keys
	- scheduling tied with combos
- Steno
	- key events are propagated immediately 
	- but produces Plover instead of normal keyboard HID
- Caps Word
	- key events are propagated immediately 
	- manipulate modifiers
	- timeout
- Fork
	- key events are propagated immediately 
	- manipulate modifiers
	- replaces key press/event
- Combos
	- key events are delayed (via buffer_keys)
		- cancel key press/event
		- send hid/event
	- scheduling tied with Morse
- Macros
	- key events are propagated immediately 
	- send hid/event

I have likely missed some things, everything is so tied together  

For reference QMK is physical key press -> chain of hardcoded calls (each call can stop the chain) -> hid. I've skipped some things that are not relevant for this discussion, but this is mostly the whole api:
- register custom key code
- execute code on keyboard init (`keyboard_pre_init_user`, `keyboard_post_init_user`)
- hook in into the chain (`post_process_record_user`, `process_record_user`, `pre_process_record_user`). 
- hook in into the layer change (`default_layer_state_set_user`, `layer_state_set_user`).
- register/unregister keycode/modifiers
- execute code after some time (`defer_exec`, `extend_deferred_exec`, `cancel_deferred_exec`)

Question is how far do we want to deviate from this and maybe more specifically how hard do we want to lean in into the events system. Current usage of events is just observing. But if we would use to for intercepting and dispatching, would the latency be a problem? I don't have much experience with embedded in this regard 

The API I would want is for Keyboard Runnable to only have bare bones key processing logic (basically just keys and layers) and extension api. Everything else is just a collection of functionalities that user would want, that are passed to the Keyboard Runnable during instantiation.

There 2 obvious ways to do it: chain of handlers (which is basically just a QMK approach, but we already deviated from it with events) or lean in into events (with custom event types produced/consumed by Keyboard runnable and interceptors). Main question here is what to do with ordering of processing and how much to allow user to change that ordering? QMK handles this by hardcoding whole processing chain, which I personally would hope we would not do. But some functionality in RMK is dependent on order of what will be produced. One case popped up at me during research is modifiers resolution would be different based on order of processing of fork, osm and capsword. But to make it more complicated I feel that the functionality should be enablable/disablable by the user. i.e. if user doesn't use the functionality they should be able to remove it, making development/debugging of keyboard for their use case not interfere with the stuff they would not use

All this customization makes sense only for rust code and TOML config should always use the same preconfigured set of functionalities.

I would be interested to hear opinions

**Comment by Exidex** [comment]
If latency is a problem, events could be done via calling processing code inside the publish function synchronously instead of going through mpmc channel

**Comment by HaoboGu** [comment]
@Exidex Thanks for the write-up! I completely agree that the current key processing logic is a bit messy. It's an extremely difficult task—not just because of the coupled logic, but also because the cross-behavior interactions are quite complex, especially when combos and Morse overlaps are involved.

I actually have some ideas in my head, such as creating multiple local channels (`Channel<NoopRawMutex, ..>`) and processing 1 event per loop. I think this is the right direction since it aligns closely with the current event system, and the overhead of a local channel with a no-op raw mutex is practically zero. If we can implement this, intercepting events and customizing behaviors will become much easier.
